### ISSUE TITLE
stack build --skip doesn't skip anything

### ISSUE THREAD

**Issue by wiz** [commercialhaskell/stack#3827]
### Steps to reproduce

```
mkdir stackholder
cd stackholder
stack new this
stack new that
rm {this,that}/stack.yaml
stack init
```

stack build --skip this
stack build --test --skip this

### Expected

Package `this` does not gets built and its tests are skipped.

### Actual

Package `this` is building and then its tests are running.

### Stack version

```
$ stack --version
Version 1.6.3, Git revision b27e629b8c4ce369e3b8273f04db193b060000db (5454 commits) x86_64 hpack-0.20.0
```

### Method of installation

* Official binary, downloaded from stackage.org or fpcomplete's package repository

**Comment by mgsloan** [comment]
Hmm, indeed, this appears to be a bug.  Original implementation is https://github.com/commercialhaskell/stack/pull/3148/files#diff-d43656b6dccca5b1708eaf03ea8bd1daR187

Looks like everything but libraries can be filtered with `--skip` currently.

**Comment by vanceism7** [comment]
I can do this one

**Comment by dbaynard** [comment]
OK — @qrilka is working on a complete rewrite of how stack works out what to build; @qrilka this shouldn't interfere, right?

**Comment by qrilka** [comment]
CLI args parsing logic wasn't supposed to be touched by refactoring in #4412 but maybe that PR could give some merge conflicts as some functions got moved a bit.

**Comment by dbaynard** [comment]
OK — I was thinking that skipping the library may have some knock on impact, but I think I'm misunderstanding the `--skip` flag's purpose.

@vanceism7 Go for it and we can deal with #4421 after it's working on `master`.

**Comment by vanceism7** [comment]
Ok, so I think I've basically got skip for libraries working, but just to check, what exactly is the behavior for the skip flag supposed to be?

With my fix in place, currently what I'm seeing based on the original issue's steps to reproduce, is that calling `stack build --skip this` would not skip the `this` package anyways since by default `stack new` creates packages that are both libs and exes. 

In order to skip the `this` package, I need to supply two skip flags, e.g: `stack build --skip this --skip this-exe`. 
Also, when running `stack test`, I need to supply three skip flags `this`, `this-exe`, and `this-test`. Is this the intended behavior?

**Comment by dbaynard** [comment]
Yes — `--skip` (should) skip a component. So in addition, if you have multiple libraries/executables/test suites you should be able to skip them individually. e.g. Two test suites (`light`, `very-long-integration-test`), run `stack test --skip very-long-integration-test` should build the library, the executable, the test suite `light` and run those tests. It should neither build nor run tests for `very-long-integration-test`.

I'm not actually sure what the value is of not building the library, though — perhaps you wish to avoid recompilation, or run a test suite on a binary?
